cli/args/flags.rs

```diff
@@ -846,6 +847,13 @@ fn compile_subcommand<'a>() -> Command<'a> {
   runtime_args(Command::new("compile"), true, false)
     .trailing_var_arg(true)
     .arg(script_arg().required(true))
+    .arg(
+      Arg::new("side-module")
```

Comment:
It would be good to add some help text for what this argument does.

Reply:
I don't think it's appropriate to make this a command line argument. The fact that a module requires certain modules to function is information about the program that is 100% determined by the author and not by the user. So it should be specified by the author, maybe in package.json or a pragma or... they could add `#import "that-module";` somewhere else in their code.

Reply:
@aapoalas suggested something similar in https://github.com/denoland/deno/pull/17663#issuecomment-1426832462. I have since realized that unreachable dynamic imports added somewhere on the main thread's module map wouldn't really work for web workers when typechecking is on.

It seems like when a module is added as a root of the module graph, with this command line argument, it is by default typechecked with the `"deno.window"` lib, but if it includes a triple-slash directive, it will use the directive's lib. That is not ideal, but it's workable. However, when a module is reachable from the main thread's module graph, it will use the `"deno.window"` lib _regardless of any triple-slash directives_. While I haven't tested the static analysis in the new version of eszip yet, I suspect it will work like that as well.

Reply:
> However, when a module is reachable from the main thread's module graph, it will use the "deno.window" lib in addition to any triple-slash directives, resulting in duplicate type definitions. While I haven't tested the static analysis in the new version of eszip yet, I suspect it will work like that as well. Meaning that this "hack" wouldn't really work for web workers.

That seems like a detail (potentially a bug?) that can be worked out or fixed. Type check them both with and witout default lib maybe?

Reply:
Maybe a bare import statement could be done from the main module that runs in the worker? If deno spots it there it can assume it is run from a worker.

Reply:
There is no "main module that runs in the worker" since the `Worker` constructor isn't statically analysable (it can be overridden both locally and globally). This PR is essentially about providing that main worker module.

Not an easy problem to solve. On one hand you'd expect dynamic imports to be type checked, on the other hand you want to avoid them because they're probably made dynamic for a reason, f.ex. being platform specific. For the hack that I proposed, adding modules to the module graph for having them be available for Worker scripts, it's beneficial that they're not type checked but also their inclusion into the module graph must not place any type libs (`"deno.window"`) on them either...

The real answer would be the [asset reference proposal](https://github.com/tc39/proposal-asset-references) though even then the question of how to do types is a bit tough.

Reply:
I'd suggest to treat these as orthogonal problems: the fact that it is not possible for a module to declare a dependency on a worker script is a different problem that this PR doesn't need to solve. 

It is also a real problem though esp. when a "complete" eszip must be built (e.g. when deploying to deploy).

Reply:
Maybe we could have deno_graph analyze for Worker constructors, remove this flag, then have a requirement for `deno compile` that everything needs to be statically analyzable for now?

Reply:
While Worker constructs are not 100% statically analysable due to the overriding issue, I do have to admit that it's probably a good idea. Overriding the constructor is anyways a very minor case, even less so with a `compile` usage on top.

If we leave it at that, supporting statically analysable dynamic imports and Workers without providing this flag, then the end result will be that users with more complex `import() + compile` or `Worker + compile` cases will be forced to use my hacky idea of eg. including arrays of `() => import("./static-name.ts")` functions in their code to provide those modules into the module graph.

That is not really the worst situation imaginable. I could live with that. But, I also am a great fan of hacky solutions that skirt the edge between what is possible with the tools given and what is proper.

Reply:
> The real answer would be the [asset reference proposal](https://github.com/tc39/proposal-asset-references) though even then the question of how to do types is a bit tough.

At this point the more relevant proposals for this are [module expressions](https://github.com/tc39/proposal-module-expressions), [module declarations](https://github.com/tc39/proposal-module-declarations) and [import reflection](https://github.com/tc39/proposal-import-reflection). Since those are actually progressing, there is a real concern of future-proofness with starting to statically analyse `Worker` specifiers.

Is it out of the question to just... allow compiled modules to use the runtime module loader and file system cache on the executing machine, and print warnings when it ever needs to do so / error if incapable? It will always be a subpar experience when something works with `deno run` but not with `deno compile`. I don't know if it's a goal to try (again) to use a stripped down executable with compiled scripts but that seems like a maintenance burden anyway.

Reply:
@dsherret 's original point has gotten somewhat lost in this thread but was very much correct: `.help("Some help here")` should be added to explain the option.

Reply:
Done. Hopefully the long help is descriptive enough to the average user.